drivers/i3c/i3c_npcx.c

```diff
@@ -2144,41 +2921,51 @@ static int npcx_i3c_init(const struct device *dev)
 		return ret;
 	}
 
+	/* Lock initial */
 	k_mutex_init(&data->lock_mutex);
 	k_sem_init(&data->sync_sem, 0, 1);
 	k_sem_init(&data->ibi_lock_sem, 1, 1);
+	k_sem_init(&data->target_lock_sem, 1, 1);
+	k_sem_init(&data->target_event_lock_sem, 1, 1);
 
 	ret = i3c_addr_slots_init(dev);
 	if (ret != 0) {
 		LOG_ERR("Addr slots init fail %d", ret);
 		return ret;
 	}
 
-	ctrl_config->is_secondary = false; /* Currently can only act as primary controller. */
-	ctrl_config->supported_hdr = I3C_MSG_HDR_DDR; /* HDR-DDR mode is supported. */
-	ctrl_config->scl.i3c = config->clocks.i3c_pp_scl_hz; /* Set I3C frequency */
+	/* Set controller default configuration */
+	config_cntlr->supported_hdr = I3C_MSG_HDR_DDR;        /* HDR-DDR mode is supported. */
+	config_cntlr->scl.i3c = config->clocks.i3c_pp_scl_hz; /* Set I3C frequency */
 
-	ret = npcx_i3c_configure(dev, I3C_CONFIG_CONTROLLER, ctrl_config);
+	/* Initial the device as controller or target */
+	config_type = (config_target->enable | config_cntlr->is_secondary) ? I3C_CONFIG_TARGET
+									   : I3C_CONFIG_CONTROLLER;
+	ret = (config_type == I3C_CONFIG_CONTROLLER) ? npcx_i3c_cntlr_init(dev)
+						     : npcx_i3c_target_init(dev);
```

Comment:
I still do not think this will cover a valid case... but then again I do not believe handoff support is in this driver yet (or at least within the Zephyr i3c framework yet) but this will likely not work when it comes...

What I'd expect here is that there would be only a `is_secondary` prop within the dts, and you've already got the initial BCR within the device tree where bits[7:6] define if it is a controller capable target or if it is a target only device which could be read instead of having a redundant parameter of the `target-mode-en` when bits[7:6] of that BCR already defines this. If you want to make it less confusing, maybe rename `is_secondary` to be just `

![image](https://github.com/user-attachments/assets/c556f05c-b2bb-434e-bc58-f03e4caa64c1)

I would also expect both `npcx_i3c_cntlr_init` and `npcx_i3c_target_init` to be called within here.
```c
if (I3C_BCR_DEVICE_ROLE(config_target->bcr) == I3C_BCR_DEVICE_ROLE_I3C_CONTROLLER_CAPABLE) {
	if (config_cntlr->is_secondary) {
		/* Secondary Controller, so boot as a target */
		cntlr_dev_en = MCONFIG_CTRENA_CAPABLE; /* Controller mode capable */
		inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Target mode enable */
	} else {
		/* Primary Controller, so boot as a controller */
		cntlr_dev_en = MCONFIG_CTRENA_ON; /* Controller mode on */
	}
	npcx_i3c_cntlr_init(dev);
	npcx_i3c_target_init(dev);
} else {
	cntlr_dev_en = MCONFIG_CTRENA_OFF; /* Controller mode off */
	npcx_i3c_target_init(dev);
	inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Target mode enable */
}

SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, cntlr_dev_en);
```
_Along with he setting of the `MCONFIG_CTRENA_*` removed in the `npcx_i3c_cntlr_init` and `npcx_i3c_target_init`._

but then again.. maybe there's an understanding with the IP with how `MCONFIG_CTRENA_CAPABLE`, `MCONFIG_CTRENA_ON`, and `MCONFIG_CTRENA_OFF` I don't know about with how it is used. Is the datasheet public (or under NDA?). If so, where I can take a look at it?

Reply:
Couple of questions about handoff I'd aligned with
1. If npcx4 bootup as controller mode, the i3c framwrok in zephyr does not support handoff yet.
2. If bootup as target mode, the npcx currently can do the handoff through npcx_i3c_target_ibi_raise(). 
   If handoff success, it'll swith to controller in target isr (Line 2845).
3. 
```c
if (I3C_BCR_DEVICE_ROLE(config_target->bcr) == I3C_BCR_DEVICE_ROLE_I3C_CONTROLLER_CAPABLE) {
	npcx_i3c_cntlr_init(dev);
	npcx_i3c_target_init(dev);
	if (config_cntlr->is_secondary) {
		/* Secondary Controller, so boot as a target */
		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_CAPABLE);
		inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Target mode enable */
	} else {
		/* Primary Controller, so boot as a controller */
		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_ON);
	}
} else {
	npcx_i3c_target_init(dev);
	SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_OFF);
	inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Target mode enable */
}
```

Reply:
Thanks for your help. 
I'm still checking if the datasheet is public, and it might take a while.

> I still do not think this will cover a valid case... but then again I do not believe handoff support is in this driver yet (or at least within the Zephyr i3c framework yet) but this will likely not work when it comes...

Couple of questions about handoff I'd  aligned with, just in case I missed something during the implementation process.
1. For controller mode, the i3c framwrok in zephyr does not support handoff yet.
   (`I3C_IBI_CONTROLLER_ROLE_REQUEST` in `i3c_ibi_workq.c ` still in progress)
2. If bootup as target mode, the npcx currently can do the handoff through ibi_raise(). 
   If handoff success, it'll swith to controller in target isr.
    ```c
    static void npcx_i3c_target_isr(const struct device *dev)
    {
    
     .... 
	    /* Secondary controller (Controller register).
	     * Check I3C now bus controller.
	     * Disable target mode if target switch to controller mode success.
	     */
	    if (IS_BIT_SET(inst->MINTMASKED, NPCX_I3C_MINTMASKED_NOWCNTLR)) {
		    inst->MSTATUS = BIT(NPCX_I3C_MSTATUS_NOWCNTLR); /* W1C */
		    inst->CONFIG &= ~BIT(NPCX_I3C_CONFIG_TGTENA);   /* Disable target mode */
		    config_cntlr->is_secondary = false;             /* Switch to primary controller */
		    npcx_i3c_apply_cntlr_config(dev);
	    }
    
    ....
    }
    ```
    So, the current implementation does not support transferring bus control in controller mode but does support requesting bus control in target mode.

3. For the suggested init sequence with BCR, based on the code above, the target will init as controller once the controller request is success.
So I update the code as below. 
```c
if (I3C_BCR_DEVICE_ROLE(config_target->bcr) == I3C_BCR_DEVICE_ROLE_I3C_CONTROLLER_CAPABLE) {
       // npcx_i3c_apply_cntlr_config(dev);     /* Removed */
       // npcx_i3c_apply_target_config(dev);  /* Removed */  

	if (config_cntlr->is_secondary) {
                npcx_i3c_apply_target_config(dev);  /* Add Initial target here .... */
		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_CAPABLE);
		inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA);
	} else {
                npcx_i3c_apply_cntlr_config(dev); /* Add Initial controller here ....*/
		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_ON);
	}
} else {
       ....
}
```

Reply:
I don't think you quite got the terminology right here (and yes it is confusing...)... but after the active controller does a handoff with GETACCCR to itself, it becomes the new "Active Controller". Keep in mind that Active Controller is different from the terms Primary and Secondary Controller. The Primary Controller is the "1st" Active Controller on the Bus. All other devices are targets (or controller-capable targets (secondary controllers). When the handoff happens, it is still called a "secondary controller", but it is now the "active controller" of which only one can exist on an I3C bus at a time. You shouldn't need these lines below in there when it comes.
```c
config_cntlr->is_secondary = false;             /* Switch to primary controller */
npcx_i3c_apply_cntlr_config(dev);
```

Anyways... I see you made the changes, so it LGTM now. Thanks!!

Reply:
Done.
These two lines are removed.